File: src/quantized/routes/magnetometry.py
"""Thin magnetometry routes. Wraps ``calc.magnetometry`` (golden vs MATLAB):
hysteresis-loop analysis, high-T background subtraction, sample-aware unit
conversion. Validate -> call the pure fn -> serialize; no analysis here.
"""

from __future__ import annotations

from typing import Any

import numpy as np
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from quantized.calc.magnetometry import (
    convert_mag_units,
    hysteresis_analysis,
    subtract_hysteresis_background,
    subtract_mag_background,
)
from quantized.routes._payload import to_jsonable

router = APIRouter(prefix="/api/magnetometry", tags=["magnetometry"])


class HysteresisRequest(BaseModel):
    h: list[float]
    m: list[float]
    saturation_fraction: float = 0.8
    pre_smooth: int = 0
    virgin_detect: bool = True


class SubtractBgRequest(BaseModel):
    temperature: list[float]
    moment: list[float]
    fit_range: tuple[float, float] | None = None
    auto_fraction: float = 0.1


class HysteresisBgRequest(BaseModel):
    h: list[float]
    m: list[float]
    hi_fraction: float = 0.7
    min_points: int = 4


class ConvertUnitsRequest(BaseModel):
    x: list[float]
    y: list[float]
    from_field: str = "Oe"
    to_field: str = "Oe"
    from_moment: str = "emu"
    to_moment: str = "emu"
    sample_mass: float = 0.0
    sample_volume: float = 0.0


@router.post("/hysteresis")
def hysteresis(req: HysteresisRequest) -> dict[str, Any]:
    """Analyze an M-H loop -> Hc / Mr / Ms / squareness / loop area / SFD."""
    try:
        result = hysteresis_analysis(
            req.h,
            req.m,
            saturation_fraction=req.saturation_fraction,
            pre_smooth=req.pre_smooth,
            virgin_detect=req.virgin_detect,
        )
    except (ValueError, IndexError) as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return to_jsonable(result)  # type: ignore[no-any-return]


@router.post("/subtract-background")
def subtract_background(req: SubtractBgRequest) -> dict[str, Any]:
    """Subtract a linear high-T background from M(T)."""
    try:
        corrected, slope, intercept = subtract_mag_background(
            req.temperature,
            req.moment,
            fit_range=req.fit_range,
            auto_fraction=req.auto_fraction,
        )
    except (ValueError, IndexError) as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return {
        "corrected": to_jsonable(corrected),
        "slope": to_jsonable(slope),
        "intercept": to_jsonable(intercept),
    }


@router.post("/subtract-hysteresis-background")
def subtract_hysteresis_bg(req: HysteresisBgRequest) -> dict[str, Any]:
    """Subtract a linear dia/paramagnetic slope from an M-H loop (slope only;
    offset kept so Hc/Mr are unaffected)."""
    try:
        corrected, slope = subtract_hysteresis_background(
            req.h, req.m, hi_fraction=req.hi_fraction, min_points=req.min_points
        )
    except (ValueError, IndexError) as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return {"corrected": to_jsonable(corrected), "slope": to_jsonable(slope)}


@router.post("/convert-units")
def convert_units(req: ConvertUnitsRequest) -> dict[str, Any]:
    """Convert field (x) + moment (y) units (sample-aware)."""
    try:
        x_out, y_out, x_unit, y_unit, warning = convert_mag_units(
            req.x,
            req.y,
            from_field=req.from_field,
            to_field=req.to_field,
            from_moment=req.from_moment,
            to_moment=req.to_moment,
            sample_mass=req.sample_mass,
            sample_volume=req.sample_volume,
        )
    except (ValueError, IndexError) as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return {
        "x": to_jsonable(np.asarray(x_out, dtype=float)),
        "y": to_jsonable(np.asarray(y_out, dtype=float)),
        "x_unit": x_unit,
        "y_unit": y_unit,
        "warning": warning,
    }

File: frontend/src/lib/api/magnetometry.ts
// /api/magnetometry/* wrappers — split out of lib/api.ts (R8 bundle-diet
// pass, 2026-08-23; see api/reference.ts's header for why): every consumer
// is the lazy magtools/hysteresis workshops. NOT re-exported by lib/api.ts;
// they import directly from this path.

import { postJSON } from "./http";
import type { CalcResult } from "../types";

/** Analyze an M-H loop -> Hc / Mr / Ms / squareness / loop area / SFD. */
export function hysteresisAnalysis(body: {
  h: number[];
  m: number[];
  saturation_fraction?: number;
  pre_smooth?: number;
  virgin_detect?: boolean;
}): Promise<CalcResult> {
  return postJSON("/api/magnetometry/hysteresis", body);
}

/** Subtract a linear high-T background from M(T) -> corrected moment + fit.
 *  M-vs-T ONLY: the window is one-sided, so on a hysteresis loop the intercept
 *  it removes carries +Ms and the loop is sheared down by that much. Use
 *  {@link subtractHysteresisBackground} for M(H) (BUG-021). Both routes take
 *  pydantic `list[float]`, which rejects the `null` a NaN stringifies to —
 *  drop gaps with `lib/api/finitePairs.ts` before calling. */
export function subtractMagBackground(body: {
  temperature: number[];
  moment: number[];
  fit_range?: [number, number] | null;
  auto_fraction?: number;
}): Promise<{ corrected: (number | null)[]; slope: number; intercept: number }> {
  return postJSON("/api/magnetometry/subtract-background", body);
}

/** Remove a linear dia/paramagnetic background from an M(H) loop: both
 *  saturated tails are fit separately, the average slope (susceptibility χ) is
 *  removed, and the loop is re-centred on its saturation midpoint. The third
 *  returned quantity is an `offset` — the vertical shift removed — NOT the
 *  intercept `subtractMagBackground` reports. */
export function subtractHysteresisBackground(body: {
  h: number[];
  m: number[];
  hi_fraction?: number;
  min_points?: number;
}): Promise<{ corrected: (number | null)[]; slope: number; offset: number }> {
  return postJSON("/api/magnetometry/subtract-hysteresis-background", body);
}

/** Convert field (x) + moment (y) units, sample-aware (emu→emu/g needs mass). */
export function convertMagUnits(body: {
  x: number[];
  y: number[];
  from_field?: string;
  to_field?: string;
  from_moment?: string;
  to_moment?: string;
  sample_mass?: number;
  sample_volume?: number;
}): Promise<{
  x: (number | null)[];
  y: (number | null)[];
  x_unit: string;
  y_unit: string;
  warning: string;
}> {
  return postJSON("/api/magnetometry/convert-units", body);
}
